pkg/github/client.go

```diff
@@ -0,0 +1,126 @@
+
+	invokeThreeTimesIfNotSuccess(func() *gogh.Response {
+		var response *gogh.Response
+		permissionLevel, response, err = c.Client.Repositories.GetPermissionLevel(context.Background(), owner, repo, user)
+		return response
+	})
+
+	return permissionLevel, err
+}
+
+// GetPullRequest retrieves information about a single pull request.
+func (c *Client) GetPullRequest(owner, repo string, prNumber int) (*gogh.PullRequest, error) {
+	var pr *gogh.PullRequest
+	var err error
+
+	invokeThreeTimesIfNotSuccess(func() *gogh.Response {
+		var response *gogh.Response
+		pr, response, err = c.Client.PullRequests.Get(context.Background(), owner, repo, prNumber)
+		return response
+	})
+
+	return pr, err
+}
+
+// ListPullRequestFiles lists the changed files in a pull request.
+func (c *Client) ListPullRequestFiles(owner, repo string, prNumber int) ([]scm.ChangedFile, error) {
+	var files []*gogh.CommitFile
+	var err error
+
+	invokeThreeTimesIfNotSuccess(func() *gogh.Response {
+		var response *gogh.Response
+		files, response, err = c.Client.PullRequests.ListFiles(context.Background(), owner, repo, prNumber, nil)
+		return response
+	})
+
+	changedFiles := make([]scm.ChangedFile, 0, len(files))
+	for _, file := range files {
+		changedFiles = append(changedFiles, scm.ChangedFile{Name: *file.Filename, Status: *file.Status})
+	}
+
+	return changedFiles, err
+}
+
+// ListIssueComments lists all comments on the specified issue.
+func (c *Client) ListIssueComments(issue scm.RepositoryIssue) ([]*gogh.IssueComment, error) {
+	var comments []*gogh.IssueComment
+	var err error
+
+	invokeThreeTimesIfNotSuccess(func() *gogh.Response {
+		var response *gogh.Response
+		comments, response, err =
+			c.Client.Issues.ListComments(context.Background(), issue.Owner, issue.RepoName, issue.Number, &gogh.IssueListCommentsOptions{})
+		return response
+	})
+
+	return comments, err
+}
+
+// CreateIssueComment creates a new comment on the specified issue.
+func (c *Client) CreateIssueComment(issue scm.RepositoryIssue, commentMsg *string) error {
+	var err error
+	comment := &gogh.IssueComment{
+		Body: commentMsg,
+	}
+
+	invokeThreeTimesIfNotSuccess(func() *gogh.Response {
+		var response *gogh.Response
+		_, response, err = c.Client.Issues.CreateComment(context.Background(), issue.Owner, issue.RepoName, issue.Number, comment)
+		return response
+	})
+
+	return err
+}
+
+// CreateStatus creates a new status for a repository at the specified reference represented by a RepositoryChange
+func (c *Client) CreateStatus(change scm.RepositoryChange, repoStatus *gogh.RepoStatus) error {
+	var err error
+
+	invokeThreeTimesIfNotSuccess(func() *gogh.Response {
+		var response *gogh.Response
+		_, response, err =
+			c.Client.Repositories.CreateStatus(context.Background(), change.Owner, change.RepoName, change.Hash, repoStatus)
+		return response
+	})
+
+	return err
+}
+
+type requestInvoker func() *gogh.Response
+
+func invokeThreeTimesIfNotSuccess(invokeRequest requestInvoker) {
```

Comment:
Since we are on re-tries I would think about making it general purpose.

```golang
// amount of retries, sleep before another one, function to call
retry.Do(3, 1 * time.Second, func() (err error) {
  // actual call
}
```

and then move it away to its own package or maybe re-use existing library (even though it sounds like a rather easy piece of code to write).

WDYT?

Reply:
Then retries and sleep can be part of the `Client` struct and we could make it configurable ease in the `plugin_init` later if needed.

Reply:
We might also want to log if we fail after all retries.

Reply:
> Since we are on re-tries I would think about making it general purpose. and then move it away to its own package

good idea, I'll move it to `util` directory and set the package to `retry`, would that work?

What do you think that should be the default timeout for GH client? 1 sec?

> re-use existing library (even though it sounds like a rather easy piece of code to write).

I took a look at some libraries, but those are focused (developed) for the purpose of using it directly with the http requests - our use-case is slightly different - we don't provide either url nor the request, but a function that returns a response. I didn't spend much time with looking for such a library (or piece of an existing code) as it seemed to be very easy, so coding it took me less time than looking for the libraries

> Then retries and sleep can be part of the Client struct and we could make it configurable ease in the plugin_init later if needed.

+1

> We might also want to log if we fail after all retries.

I was thinking about it, but as we log all the errors (if necessary) in the callers, then it didn't seem to be important to log it here as well. However, log it with the `DEBUG` level could be fine.

Reply:
> What do you think that should be the default timeout for GH client? 1 sec?

Yeah, that is probably good starting point

> I took a look at some libraries

I look at some too [[1](https://github.com/avast/retry-go)][[2](https://medium.com/@matryer/retrying-in-golang-quicktip-f688d00e650a)], but I would still consider writing this on our own.

> I was thinking about it, but as we log all the errors (if necessary) in the callers, then it didn't seem to be important to log it here as well. However, log it with the `DEBUG` level could be fine.

Except that there is no `DEBUG` level ;) You are right that we log errors, but now if we reach the limit of retries what kind of error do we return to the caller? We should make sure we propagate the right one.

Reply:
first two points implemented here: 257d959
it is not possible to have multiple packages inside of one directory, so I created a new dir `http`, put the `retry.go` file there and also moved `url.go` file into this dir.
Now, I'm going to implement returning of the error